src/Linters/UseCanonicalDataProviders.hack

```diff
@@ -0,0 +1,312 @@
+    $info = $this->getInfoAboutTestFile($context);
+    if ($info is null) {
+      return null;
+    }
+
+    list($data_providers, $hhast_methods) = tuple(
+      $info['data_providers'],
+      $info['hhast_methods'],
+    );
+
+    $method_name = $node->getName()->getText();
+
+    if (!C\contains_key($data_providers, $method_name)) {
+      return null;
+    }
+
+    $hhast_method = $hhast_methods[$method_name];
+
+    list($err, $tuple_type) = $this->jugdeReturnTypes($hhast_method, $node);
+
+    if ($err is nonnull) {
+      return $err;
+    }
+
+    foreach ($data_providers[$method_name] as list($test, $_)) {
+      $err = $this->judgeTestTypes(
+        $node,
+        $tuple_type as nonnull,
+        $hhast_methods[$test],
+      );
+      if ($err is nonnull) {
+        return $err;
+      }
+    }
+
+    return null;
+  }
+
+  private function jugdeReturnTypes(
+    FunctionDeclarationHeader $hhast_method,
+    this::TNode $node,
+  ): (?ASTLintError, ?TupleTypeSpecifier) {
+    $return_type = $hhast_method->getType();
+
+    // Only in partial files
+    if ($return_type is null) {
+      return tuple(
+        new ASTLintError(
+          $this,
+          'Add a return type to your dataprovider',
+          $node,
+          () ==> null,
+        ),
+        null,
+      );
+    }
+
+    if ($return_type is VectorTypeSpecifier) {
+      $inner_type = $return_type->getTypex();
+    } else if ($return_type is DictionaryTypeSpecifier) {
+      $inner_type = $return_type->getMembersx()->toVec()[1]->getItem();
+    } else {
+      return tuple(
+        new ASTLintError(
+          $this,
+          'Use vec<_> or dict<_, _> return types',
+          $node,
+          () ==> null,
+        ),
+        null,
+      );
+    }
+
+    if (!$inner_type is TupleTypeSpecifier) {
+      return tuple(
+        new ASTLintError(
+          $this,
+          'Use a tuple as the value type of your return type',
+          $node,
+          () ==> null,
+        ),
+        null,
+      );
+    }
+
+    return tuple(null, $inner_type);
+  }
+
+  private function judgeTestTypes(
+    this::TNode $node,
+    TupleTypeSpecifier $data_provider_types,
+    FunctionDeclarationHeader $test,
+  ): ?ASTLintError {
+    $test_types_with_nulls = ($test->getParameterList()?->toVec() ?? vec[])
+      |> Vec\map($$, $li ==> $li->getItem()->getType());
+
+    $test_types = Vec\filter_nulls($test_types_with_nulls);
+
+    if (C\count($test_types_with_nulls) !== C\count($test_types)) {
+      return new ASTLintError(
+        $this,
+        Str\format('Add parameter types to %s', $test->getName()->getText()),
+        $node,
+        () ==> null,
+      );
+    }
+
+    $tuple_types = $data_provider_types->getTypes()->toVec()
+      |> Vec\map($$, $li ==> $li->getItem());
+
+    $tuple_arity = C\count($tuple_types);
+    $test_arity = C\count($test_types);
+
+    if ($tuple_arity !== $test_arity) {
+      return new ASTLintError(
+        $this,
```

Comment:
Just to check:
- this raises an error on the dataprovider for test mismatch?
- IMO it should probably be on the test function instead, given you can use a single data provider for multiple...
- ... but doing that sucks and is slow in HHAST at the moment?

Reply:
The ASTLintError constructor takes a Node argument. Am I breaking the contract of getLintErrorForNode if I blame a different node than the one being given to me? If not, I can blame the FunctionDeclHeader of the test instead.

Reply:
Hmm, yeah, it's currently a bit broken: getPrettyTextForNode takes a TNode but is called with others:

```
ASTLintError.hack
61:    return $this->linter->getPrettyTextForNode(/* HH_FIXME[4110] */$this->node);
```

The right thing to do is probably to make that take a Node, not a TNode, and make linters use `as ` etc.

The alternative fix is to make ASTLintError generic again, but that led to a lot of complicated spaghetti

Reply:
Let's stick with blaming the data provider for now.
We can make an issue about the getPrettyTextForNode() problem.
I am torn about the generic on LintError.
I'll have to weigh the up and downsides.

Reply:
Yep, very much in favor of fixing that separately later, not blocking this :)